File: backend/app/schemas/writing.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class NovelBlueprintOut(BaseModel):
    id: int
    novel_id: int
    version: int
    status: str
    content_json: dict[str, Any]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BlueprintGenerateRequest(BaseModel):
    author_input: str


class BlueprintUpdateRequest(BaseModel):
    content_json: Optional[dict[str, Any]] = None


class ChapterPlanOut(BaseModel):
    id: int
    novel_id: int
    chapter_id: Optional[int] = None
    position: int
    status: str
    content_json: dict[str, Any]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ChapterPlanUpdateRequest(BaseModel):
    content_json: Optional[dict[str, Any]] = None


class ChapterBriefOut(BaseModel):
    id: int
    novel_id: int
    chapter_plan_id: int
    chapter_id: Optional[int] = None
    status: str
    brief_json: dict[str, Any]
    length_contract_json: dict[str, Any]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ChapterBriefGenerateRequest(BaseModel):
    chapter_plan_id: int
    plot_plan_revision_id: Optional[int] = None
    author_input: Optional[str] = None


class ChapterBriefUpdateRequest(BaseModel):
    brief_json: Optional[dict[str, Any]] = None
    length_contract_json: Optional[dict[str, Any]] = None


class ContextPackageGenerateRequest(BaseModel):
    chapter_brief_id: int
    plot_plan_revision_id: Optional[int] = None
    author_input: Optional[str] = None


class ContextPackageOut(BaseModel):
    id: int
    novel_id: int
    chapter_brief_id: int
    package_json: dict[str, Any]
    created_at: datetime

    model_config = {"from_attributes": True}


class WritingRunOut(BaseModel):
    id: int
    novel_id: int
    chapter_brief_id: int
    context_package_id: int
    target_chapter_id: Optional[int] = None
    status: str
    draft_content: str
    word_count: int
    gate_result_json: dict[str, Any]
    gated: bool = False
    has_pending_repairs: bool = False
    error_message: Optional[str] = None
    accepted_at: Optional[datetime] = None
    decision_id: Optional[int] = None
    context_snapshot_json: dict[str, Any] = {}
    planning_blocked: bool = False
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class WritingRunCreateRequest(BaseModel):
    chapter_brief_id: int
    context_package_id: Optional[int] = None
    plot_plan_revision_id: Optional[int] = None
    author_input: Optional[str] = None


class RepairLogOut(BaseModel):
    id: int
    novel_id: int
    writing_run_id: int
    issue_type: str
    description: str
    location: str
    old_text: str
    new_text: str
    created_at: datetime

    model_config = {"from_attributes": True}


class PendingRepairOut(BaseModel):
    id: int
    novel_id: int
    chapter_id: Optional[int] = None
    writing_run_id: int
    issue_type: str
    description: str
    location: str
    context: str
    options: Optional[list] = None
    intent_type: str
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class ResolveRepairRequest(BaseModel):
    action: str = Field(...)  # "apply"
    choice_index: Optional[int] = None
    intent_text: Optional[str] = None


class RepairsResponse(BaseModel):
    repair_logs: list[RepairLogOut]
    pending_repairs: list[PendingRepairOut]

File: frontend/src/types/writing.ts
export interface NovelBlueprint {
  id: number
  novel_id: number
  version: number
  status: 'draft' | 'active' | 'archived'
  content_json: {
    core_promise: string
    theme: string
    main_conflict: string
    character_arcs: string
    world_rules: string
    narrative_perspective: string
    style_constraints: string
    ending_direction: string
  }
  created_at: string
  updated_at: string
}

export interface ChapterPlan {
  id: number
  novel_id: number
  chapter_id: number | null
  position: number
  status: 'draft' | 'ready' | 'used'
  content_json: {
    chapter_title: string
    plot_task: string
    character_task: string
    information_task: string
    emotional_effect: string
    pacing: string
    foreshadowing_task: string
  }
  created_at: string
  updated_at: string
}

export interface ChapterBrief {
  id: number
  novel_id: number
  chapter_plan_id: number
  chapter_id: number | null
  status: 'draft' | 'ready' | 'used'
  brief_json: {
    writing_goal: string
    scenes: { name: string; purpose: string; conflict: string; expected_words: number }[]
    participating_characters: string
    conflict_design: string
    information_control: string
    foreshadowing_handling: string
    writing_constraints: string
    acceptance_criteria: string
  }
  length_contract_json: {
    target_words: number
    min_words: number
    max_words: number
    scene_word_allocation: { scene: string; words: number }[]
    density_requirements: string
    expansion_strategy: string
    compression_strategy: string
  }
  created_at: string
  updated_at: string
}

export interface ContextPackage {
  id: number
  novel_id: number
  chapter_brief_id: number
  package_json: Record<string, unknown>
  created_at: string
}

export interface WritingRun {
  id: number
  novel_id: number
  chapter_brief_id: number
  context_package_id: number
  target_chapter_id: number | null
  status: 'running' | 'completed' | 'decision_required' | 'failed' | 'accepted' | 'discarded'
  draft_content: string
  word_count: number
  gate_result_json: {
    passed: boolean
    reasons: string[]
    min_words: number
    target_words: number
    max_words: number
    outline_like: boolean
  }
  gated: boolean
  has_pending_repairs: boolean
  error_message: string | null
  decision_id: number | null
  context_snapshot_json: Record<string, unknown>
  planning_blocked: boolean
  accepted_at: string | null
  created_at: string
  updated_at: string
}

export interface RepairLog {
  id: number
  novel_id: number
  writing_run_id: number
  issue_type: string
  description: string
  location: string
  old_text: string
  new_text: string
  created_at: string
}

export interface PendingRepairOption {
  label: string
  summary: string
}

export interface PendingRepair {
  id: number
  novel_id: number
  chapter_id: number | null
  writing_run_id: number
  issue_type: string
  description: string
  location: string
  context: string
  options: PendingRepairOption[] | null
  intent_type: 'choice' | 'freeform'
  status: 'pending' | 'applied' | 'dismissed'
  created_at: string
}

export interface RepairsResponse {
  repair_logs: RepairLog[]
  pending_repairs: PendingRepair[]
}
